processor/redactionprocessor/processor.go

```diff
@@ -0,0 +1,121 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redactionprocessor
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+)
+
+var _ component.TracesProcessor = (*redaction)(nil)
+
+type redaction struct {
+	// Attribute keys allowed in a span
+	allowList map[string]string
+	// Attribute values blocked in a span
+	blockRegexList map[string]*regexp.Regexp
+	// Redaction processor configuration
+	config *Config
+	// Logger
+	logger *zap.Logger
+	// Next trace consumer in line
+	next consumer.Traces
+}
+
+// newRedaction creates a new instance of the redaction processor
+func newRedaction(ctx context.Context, config *Config, logger *zap.Logger, next consumer.Traces) (*redaction, error) {
+	allowList := makeAllowList(config)
+	blockRegexList, err := makeBlockRegexList(ctx, config)
+	if err != nil {
+		// TODO: Placeholder for an error metric in the next PR
+		return nil, fmt.Errorf("failed to process block list: %w", err)
+	}
+
+	return &redaction{
+		allowList:      allowList,
+		blockRegexList: blockRegexList,
+		config:         config,
+		logger:         logger,
+		next:           next,
+	}, nil
+}
+
+// processTraces implements ProcessMetricsFunc. It processes the incoming data
+// and returns the data to be sent to the next component
+func (s *redaction) processTraces(_ context.Context, batch pdata.Traces) (pdata.Traces, error) {
+	// TODO: Implementation to follow in the next PR
+	return batch, nil
+}
+
+// ConsumeTraces implements the SpanProcessor interface
+func (s *redaction) ConsumeTraces(_ context.Context, _ pdata.Traces) error {
+	// TODO: Implementation to follow in the next PR
+	return nil
+}
+
+const (
+	redactedKeys     = "redacted_keys"
+	redactedKeyCount = "redacted_key_count"
+	maskedValues     = "masked_values"
+	maskedValueCount = "masked_value_count"
+)
+
+// makeAllowList sets up a lookup table of allowed span attribute keys
+func makeAllowList(c *Config) map[string]string {
+	redactionKeys := []string{redactedKeys, redactedKeyCount, maskedValues, maskedValueCount}
```

Comment:
@dmitryax Using this processor in production, I've found that the counts are convenient for querying through traces for redaction. The lists of attributes are nice for diagnostics, but being able to query for `redactedKeysCount > 0` is convenient

The design reason for the counts is that they leak less information than the lists of span attribute keys in case the actual key names need to be kept confidential

Reply:
Hey @leonsp-ai, sorry for the delayed response. Thanks for explaining the attributes purpose. It looks good to me.

Also attribute names in this PR don't fit established attribute naming conventions. Please take a look at [this specification](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) and think about names that would better fit the tracing model. Once the processor is implemented (or closer to final PRs), they should be added to the semantic conventions in the spec. Not necessary to change it in this PR, can be changed later

Reply:
@dmitryax Thanks for the link to the spec! I've pencilled in names that match a little better